src/test/java/com/puppycrawl/tools/checkstyle/utils/ChainedPropertyUtilTest.java

```diff
@@ -0,0 +1,136 @@
+import com.puppycrawl.tools.checkstyle.PropertiesExpander;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+
+public class ChainedPropertyUtilTest extends AbstractModuleTestSupport {
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/puppycrawl/tools/checkstyle/utils/chainedpropertyutil";
+    }
+
+    @Test
+    public void testIsProperUtilsClass() throws ReflectiveOperationException {
+        assertWithMessage("Constructor is not private.")
+            .that(isUtilsClassHasPrivateConstructor(ChainedPropertyUtil.class, true))
+            .isTrue();
+    }
+
+    @Test
+    public void testPropertyChaining() throws Exception {
+        final File propertiesFile =
+            new File(getPath("InputChainedPropertyUtil.properties"));
+        final Properties properties = loadProperties(propertiesFile);
+        final Properties resolvedProperties =
+            ChainedPropertyUtil.getResolvedProperties(properties);
+        final PropertiesExpander expander = new PropertiesExpander(resolvedProperties);
+        final String message = "Unexpected property resolution.";
+
+        assertWithMessage(message)
+            .that(expander.resolve("basedir"))
+            .isEqualTo("/home");
+        assertWithMessage(message)
+            .that(expander.resolve("checkstyle.dir"))
+            .isEqualTo("/home/checkstyle");
+        assertWithMessage(message)
+            .that(expander.resolve("config.dir"))
+            .isEqualTo("/home/checkstyle/configs");
+        assertWithMessage(message)
+            .that(expander.resolve("checkstyle.suppressions.file"))
+            .isEqualTo("/home/checkstyle/configs/suppressions.xml");
+        assertWithMessage(message)
+            .that(expander.resolve("checkstyle.dir"))
+            .isEqualTo("/home/checkstyle");
+    }
+
+    @Test
+    public void testPropertyChainingPropertyNotFound() throws Exception {
+        final File propertiesFile =
+            new File(getPath("InputChainedPropertyUtilUndefinedProperty.properties"));
+        final Properties properties = loadProperties(propertiesFile);
+        final String expected =
+            ChainedPropertyUtil.UNDEFINED_PROPERTY_MESSAGE + "[property.not.found]";
+        final String message = "Undefined property reference expected.";
+
+        final CheckstyleException exception =
+            assertThrows(CheckstyleException.class,
+                () -> ChainedPropertyUtil.getResolvedProperties(properties));
+
+        assertWithMessage(message)
+            .that(exception)
+            .hasMessageThat()
+            .isEqualTo(expected);
+    }
+
+    @Test
+    public void testPropertyChainingRecursiveUnresolvable() throws Exception {
+        final File propertiesFile =
+            new File(getPath("InputChainedPropertyUtilRecursiveUnresolvable.properties"));
+        final Properties properties = loadProperties(propertiesFile);
```

Comment:
This is strange.
We put test input in separate file, create separate method to load it.
Can we inline properties as string
And load properties from string https://stackoverflow.com/a/9672376/1015848
In this case we do not need jump between files to validate junit correctness.
All will be in junit method, easy to read.

It is not related for CLI testing cases.

Reply:
> Can we inline properties as string

I have no problem with this, but I would like to point out that I have noticed that property loading order from file is random.  Inlining these strings and manually loading them will make sure that order is maintained.  I thought it would be good to make sure that order is random for testing by loading from file, since this is how it will be in actual usage.

Reply:
Ok, please place this reason as comment above code that loads them, it is not obvious.
Let's keep  them in files.

Reply:
Added details to loading method javadoc.